### PR TITLE
feat: organization_token grant

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

- implement `urn:logto:grant-type:organization_token` grant per [RFC 0001](https://github.com/logto-io/rfcs).
  - a notable change is the use of reserved resource server for organization. this will be reflected to the RFC later.
  - now two things are required for the auth request: organization resource and organization scope.
  - the organization resource server (`urn:logto:resource:organizations`) is used for keeping organization template scopes in the refresh token.
  - the organization scope (`urn:logto:scope:organizations`) is used for the `organizations` claim in the ID token and authorize the new grant.
- since the grant flow is similar to the `refresh_token` grant, code is edited from the [oidc-provider implementation](https://github.com/panva/node-oidc-provider/blob/cf2069cbb31a6a855876e95157372d25dde2511c/lib/actions/grants/refresh_token.js) with necessary type definitions added for internal helpers. reviewers should check those types as well since they were derived from the js files.
- refactor the `getResourceServerInfo` function to support reserved resources.

### To-dos

- add comprehensive test cases for the new grant
- consider requiring the presence of the organization resource server for issuing access tokens
- update RFC accordingly

## Testing

performed some manual tests for the grant:

#### 1. `scope` is omitted, refresh token rotation is on

image

the response looks like:

```json
{
    "access_token": "<redacted>",
    "expires_in": 3600,
    "refresh_token": "<redacted>",
    "scope": "read:appointment delete:appointment",
    "token_type": "Bearer"
}
```

where the scopes are the intersection of initial scopes that passed to the auth request and the available scopes for the user in the organization.

the decoded access token looks like:

image

#### 2. `scope` is not omitted

Logto successfully "downscope" the token to what the scope parameter includes.

#### 3. the organization context

Logto can issue proper scopes accordingly to the user roles in the specific organization requested.

#### 4. sad paths

handled decently:

- not a member
- no sufficient scope in refresh token 

## Checklist

- [ ] `.changeset` (another PR)
- [ ] unit tests (another PR)
- [ ] integration tests (another PR)
- [x] necessary TSDoc comments

### CODE REVIEW

File: packages/core/src/event-listeners/grant.ts [hunk 1]
```diff
@@ -69,6 +69,7 @@ const grantTypeToExchangeByType: Record<GrantType, token.ExchangeByType> = {
   [GrantType.AuthorizationCode]: token.ExchangeByType.AuthorizationCode,
   [GrantType.RefreshToken]: token.ExchangeByType.RefreshToken,
   [GrantType.ClientCredentials]: token.ExchangeByType.ClientCredentials,
+  [GrantType.OrganizationToken]: token.ExchangeByType.RefreshToken,
```
Comment (wangsijie) [inline packages/core/src/event-listeners/grant.ts:72]: Could you add a comment here? It is confusing that "OrganizationToken" key with value "RefreshToken"

**Review by wangsijie — APPROVED** [review]

**Review by darcyYe — APPROVED** [review]
Overall LGTM

**Review by simeng-li — APPROVED** [review]
LGTM

File: packages/core/src/oidc/grants/organization-token.ts [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+  'refresh_token',
+  'organization_id',
+] as const) satisfies ReadonlyArray<(typeof parameters)[number]>;
+
+/**
+ * The required scope for the `urn:logto:grant-type:organization_token` grant type.
+ *
+ * @see {@link GrantType.OrganizationToken}
+ */
+const requiredScope = UserScope.Organizations;
```
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/core/src/oidc/grants/organization-token.ts:66]: this is removed in #4854

File: packages/core/src/include.d/oidc-provider/lib-keep/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+The original folder named `lib` which will be ignored by our build system. Change the name to `lib.keep` to avoid confusion.
```
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/core/src/include.d/oidc-provider/lib-keep/README.md:1]: updated in #4854

**Pull request merged**